Avoid redundant computations during central moments calculation

grlee77 opened the pull request:
## Description

The current calculation of central moments is inefficient in two ways:

1.) Raw first order moments are calculated using `moments_central` with ``center = (0, 0)`` and then then calculations are repeated using the centroid calculated in the first step. In this PR, central moments are computed directly from the raw moments without having to traverse the input mage again. This also allows exact values of 0 to be given for the first order central moments which was not the case previously due to numerical error.

2.) Some moments that have higher order than requested get generated. As a concrete example for order=2 in 2D, a fully populated 3x3 moments array is returned but only six of the nine entries actually have order <= 2. The ones at at index (1, 2) and (2, 1) are order 3 and the one at (2, 2) is order 4. The way the code is currently written doesn't allow skipping computation of these, though.

This PR resolves the first of these two issues, but does not fully address the second one (the extra raw moments still get computed, we just don't generate their corresponding central moments).

In practice, times to compute central moments improve by a factor of somewhat less than two. Other `regionprops` like inertia tensor eigenvalues and major axis length rely on computations of central moments and thus will also benefit.

A SymPy-based utility script that was used to compute the analytical terms is included in the `tools` folder of the repository. Only expressions for ndim 2 and 3 and order < 4 were retained. One could use the general expression in all cases, but that has a performance penalty over the simplified expressions since SymPy ends up being able to cancel out many of the terms during simplification.

## Checklist

<!-- It's fine to submit PRs which are a work in progress! -->
<!-- But before they are merged, all PRs should provide: -->
- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- Descriptive commit messages (see below)

<!-- For detailed information on these and other aspects see -->
<!-- the scikit-image contribution guidelines. -->
<!-- https://scikit-image.org/docs/dev/contribute.html -->

## For reviewers

<!-- Don't remove the checklist below. -->
- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

grlee77 commented:
## current main branch

```
benchmarks_moments.MomentsSuite.time_moments_central

--                                                order                 
--------------------------------- --------------------------------------
      shape            dtype           1            2            3      
================= =============== ============ ============ ============
     (64, 64)       numpy.uint8    62.8±0.5μs   62.7±0.4μs   62.9±0.2μs 
     (64, 64)      numpy.float32   51.0±0.4μs   51.1±0.2μs   51.1±0.4μs 
     (64, 64)      numpy.float64   58.7±0.3μs   58.4±0.4μs   59.0±0.4μs 
   (4096, 2048)     numpy.uint8     47.9±2ms     46.7±2ms     46.9±1ms  
   (4096, 2048)    numpy.float32   7.46±0.6ms   7.17±0.9ms   7.62±0.8ms 
   (4096, 2048)    numpy.float64    19.8±3ms     16.0±3ms     18.4±3ms  
   (32, 32, 32)     numpy.uint8     371±4μs      373±4μs      367±30μs  
   (32, 32, 32)    numpy.float32    301±3μs      303±1μs      304±6μs   
   (32, 32, 32)    numpy.float64    348±2μs      349±2μs      349±10μs  
 (256, 256, 192)    numpy.uint8     970±10ms     945±10ms     951±20ms  
 (256, 256, 192)   numpy.float32    421±10ms     429±6ms      421±10ms  
 (256, 256, 192)   numpy.float64    912±10ms     903±40ms     934±30ms  
================= =============== ============ ============ ============
```

## PR branch

```
benchmarks_moments.MomentsSuite.time_moments_central

--                                                order                 
--------------------------------- --------------------------------------
      shape            dtype           1            2            3      
================= =============== ============ ============ ============
     (64, 64)       numpy.uint8    42.0±0.2μs   43.4±0.6μs   42.0±0.2μs 
     (64, 64)      numpy.float32   36.0±0.5μs   35.8±0.2μs   35.9±0.4μs 
     (64, 64)      numpy.float64   39.9±0.6μs   58.6±20μs    40.3±0.6μs 
   (4096, 2048)     numpy.uint8     25.6±7ms     25.9±4ms     27.6±2ms  
   (4096, 2048)    numpy.float32    6.02±2ms     6.20±2ms    5.40±0.5ms 
   (4096, 2048)    numpy.float64    9.70±1ms     10.7±2ms     9.02±1ms  
   (32, 32, 32)     numpy.uint8     259±4μs      259±3μs      256±3μs   
   (32, 32, 32)    numpy.float32    212±3μs     211±0.5μs    210±0.7μs  
   (32, 32, 32)    numpy.float64    357±30μs    244±0.7μs     243±2μs   
 (256, 256, 192)    numpy.uint8     602±30ms     589±10ms     593±7ms   
 (256, 256, 192)   numpy.float32    279±6ms      276±6ms      278±5ms   
 (256, 256, 192)   numpy.float64    589±10ms     590±10ms     588±20ms  
================= =============== ============ ============ ============
```

grlee77 commented:
One caveat is potential considerations due to numerical stability. Using a two-pass approach is likely more accurate in general, so perhaps we should still provide an option to use that approach on request (or keep it as the default).

See, for example the following publication, which I have not yet reviewed in any detail: 
```Pébay, P., Terriberry, T.B., Kolla, H. et al. Numerically stable, scalable formulas for parallel and online computation of higher-order multivariate central moments with arbitrary weights. Comput Stat 31, 1305–1325 (2016).```
https://doi.org/10.1007/s00180-015-0637-z
(also available here: https://www.osti.gov/pages/biblio/1427275&usg=AOvVaw2nVh6r-3wL8v0stUs2RUa_)

Accuracy is likely a concern with either approach for high order moments, but I think up to 2nd or 3rd order moments are sufficient in most cases for purposes of scikit-image.

pep8speaks commented:
Hello @grlee77! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`benchmarks/benchmark_measure.py`](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/benchmarks/benchmark_measure.py):

> [Line 63:11](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/benchmarks/benchmark_measure.py#L63): [E124](https://duckduckgo.com/?q=pep8%20E124) closing bracket does not match visual indentation



* In the file [`skimage/measure/_moments_analytical.py`](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py):

> [Line 48:49](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L48): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 49:49](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L49): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 50:49](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L50): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 53:49](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L53): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 53:52](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L53): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 53:61](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L53): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 54:40](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L54): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 54:56](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L54): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 54:65](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L54): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 55:40](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L55): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 55:43](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L55): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 56:49](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L56): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 56:52](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L56): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 56:61](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L56): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 57:40](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L57): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 57:52](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L57): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 57:55](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L57): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 57:58](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L57): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 58:49](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L58): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 58:52](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L58): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 58:61](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L58): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 59:39](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L59): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 59:45](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L59): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 60:49](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L60): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 60:52](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L60): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 60:61](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L60): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 61:39](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L61): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 61:45](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L61): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 71:43](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L71): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 72:43](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L72): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 73:43](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L73): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 74:43](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L74): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 75:43](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L75): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 76:43](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L76): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 79:42](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L79): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 79:48](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L79): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 79:60](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L79): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 80:42](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L80): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 80:45](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L80): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 80:57](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L80): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 82:44](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L82): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 82:56](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L82): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 83:42](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L83): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 83:45](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L83): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 83:50](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L83): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 83:76](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L83): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 85:46](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L85): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 85:61](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L85): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 85:64](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L85): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 85:76](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L85): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 86:43](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L86): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 86:48](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L86): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 86:74](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L86): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 88:42](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L88): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 88:48](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L88): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 88:60](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L88): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 89:42](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L89): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 89:45](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L89): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 89:57](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L89): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 91:44](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L91): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 91:56](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L91): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 92:42](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L92): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 92:45](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L92): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 92:50](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L92): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 92:76](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L92): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 94:44](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L94): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 94:56](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L94): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 95:43](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L95): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 95:48](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L95): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 95:74](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L95): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 96:43](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L96): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 96:48](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L96): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 96:74](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L96): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 98:44](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L98): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 98:56](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L98): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 99:42](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L99): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 99:45](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L99): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 99:50](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L99): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 99:76](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L99): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 101:46](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L101): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 101:58](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L101): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 102:42](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L102): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 102:45](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L102): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 102:57](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L102): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 103:43](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L103): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 103:48](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L103): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 103:74](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L103): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 105:46](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L105): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 105:58](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L105): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 106:42](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L106): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 106:45](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L106): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 106:57](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L106): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 107:43](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L107): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 107:48](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L107): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 107:74](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L107): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 109:42](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L109): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 109:48](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L109): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 109:60](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L109): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 110:42](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L110): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 110:45](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L110): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 110:57](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L110): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 125:63](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L125): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 140:55](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L140): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 140:61](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/skimage/measure/_moments_analytical.py#L140): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator



* In the file [`tools/precompute/moments_sympy.py`](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/tools/precompute/moments_sympy.py):

> [Line 38:53](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/tools/precompute/moments_sympy.py#L38): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 39:53](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/tools/precompute/moments_sympy.py#L39): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 46:37](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/tools/precompute/moments_sympy.py#L46): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 47:37](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/tools/precompute/moments_sympy.py#L47): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 80:74](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/tools/precompute/moments_sympy.py#L80): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 81:74](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/tools/precompute/moments_sympy.py#L81): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 88:44](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/tools/precompute/moments_sympy.py#L88): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 89:44](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/tools/precompute/moments_sympy.py#L89): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 90:44](https://github.com/scikit-image/scikit-image/blob/61c32be7e9ad94b709d15895392ee32180811d34/tools/precompute/moments_sympy.py#L90): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator

mkcor commented on skimage/measure/_moments_analytical.py:
```diff
@@ -0,0 +1,157 @@
+    moments_central = np.zeros_like(moments_raw)
+    if order >= 4 or ndim not in [2, 3]:
+        raise ValueError(
+            "This function only supports 2D or 3D moments with order < 4"
```
```suggestion
            "This function only supports 2D or 3D moments of order < 4."
```

mkcor commented on skimage/measure/_moments_analytical.py:
```diff
@@ -0,0 +1,157 @@
+
+
+def _moments_raw_to_central_fast(moments_raw):
+    """Analytical equations for 2D and 3D central moments with order < 4.
```
```suggestion
    """Analytical equations for 2D and 3D central moments of order < 4.
```

mkcor commented on tools/precompute/moments_sympy.py:
```diff
@@ -0,0 +1,100 @@
+ndim = 3
+order = 3
+if ndim == 2:
+    # symbols for the centroid componets
```
```suggestion
    # symbols for the centroid components
```

mkcor commented on tools/precompute/moments_sympy.py:
```diff
@@ -0,0 +1,100 @@
+    cx = M[1, 0] / M[0, 0]
+    cy = M[0, 1] / M[0, 0]
+
+    # loop over all moments with order <= `order`.
```
```suggestion
    # Loop over all moments of order <= `order`.
```

mkcor commented on tools/precompute/moments_sympy.py:
```diff
@@ -0,0 +1,100 @@
+
+elif ndim == 3:
+
+    # symbols for the centroid componets
```
```suggestion
    # symbols for the centroid components
```

mkcor commented on tools/precompute/moments_sympy.py:
```diff
@@ -0,0 +1,100 @@
+    cz = M[0, 0, 1] / M[0, 0, 0]
+
+    # loop over all moments with order <= `order`.
+    print(f"Generating expressions for {ndim}D moments with order <= {order}")
```
```suggestion
    print(f"Generating expressions for {ndim}D moments of order <= {order}")
```

mkcor reviewed (commented):
Only had a quick look, will review thoroughly tomorrow. Thanks!

mkcor reviewed (commented):
Nice work!

It looks like `tools/precompute/moments_sympy.py` needs a little cleanup (print statements).

I have skimmed through the paper by Pébay *et al.*, and they show that the gain in efficiency obtained with their one-pass incremental approach is *not* at the expense of numerical accuracy (wrt to the 'naive' two-pass approach).

Their update formula (incrementing the centroid components in the moments matrix) is not only *incremental* but also *pairwise* though... since they partition the total set S into two sets, A and B. So I'm not sure how to compare (3.15) in the paper with our formula.

grlee77 commented on tools/precompute/moments_sympy.py:
```diff
@@ -0,0 +1,100 @@
+# Define a moments matrix, M
+M = IndexedBase('M')
+
+ndim = 3
+order = 3
```
No, this whole file is just here for future reference into how the equations in `_moments_analytical.py` were obtained. It will not be distributed with scikit-image. The idea is that a dev would just run `python moments_sympy.py` as a standalone script and it will print equations for all orders <= `order` at the specified `ndim` to the terminal. One could copy the text output from the console and paste it into `_moments_analytical.py` to reproduce expressions there.

mkcor replied:
Oh, ok! That's why you wrote "in the GitHub repository" (file `tools/precompute/moments_sympy.py` won't be in the scikit-image package). Thanks for clarifying. :pray:

grlee77 commented:
> It looks like tools/precompute/moments_sympy.py needs a little cleanup (print statements).

This was intentional (see https://github.com/scikit-image/scikit-image/pull/6188#discussion_r788213676)

alexdesiqueira reviewed (approved):
Thanks @grlee77! Looks good to me.

mkcor commented on skimage/measure/tests/test_moments.py:
```diff
@@ -173,7 +200,31 @@ def test_moments_normalized_3d():
+        x = rng.standard_normal(shape, dtype=dtype)
+    # setting center=None will use the analytical expressions
+    m1 = moments_central(x, center=None, order=order)
+    # providing explicit centroid will bypass the analytic code path
```
```suggestion
    # providing explicit centroid will bypass the analytical code path
```

mkcor reviewed (commented):
Oops! I have resolved the conflicts in a formal manner, considering the changes which landed meanwhile with #6296. Will look at the tests tomorrow or next week.

grlee77 commented:
Rebased and updated the test cases that were added in the meantime for anisotropic spacing. As for other cases, we only want to compare the moments for the upper left triangle as the elements at the lower right have moments larger than the requested order (explained in initial comment in this PR).

grlee77 commented:
@lagru, I added a 0.20 tag here in hopes we can consider this one for the release. The summary of benefits are:
- performance improvement (a bit less than a factor of 2 when I tried it)
- improved numerically accuracy

The only change in behavior is that central moments larger than the requested order are no longer computed (those elements of the matrix will be zeros). This is why some test cases use a `compare_moments` function that only compares the values in the upper left triangle of the moment matrix.

mkcor commented on skimage/measure/_moments_analytical.py:
```diff
@@ -0,0 +1,157 @@
+    # The nested loops below are an n-dimensional extension of the 2D formula
+    # given at https://en.wikipedia.org/wiki/Image_moment#Central_moments
+
+    # iterate overall [0, order] (inclusive) on each axis
```
```suggestion
    # iterate over all [0, order] (inclusive) on each axis
```

hmaarrfk replied:
Sorry to jump in here, you all are doing much better at helping maintain and improve scikit-image than I am. 

May I suggest that these stylistic nits at the end of a review be simply "applied" instead of suggested?

This would enable the merge to just go through.

The back and forth takes a lot of core developer time, and I think it would help move projects forward if these smaller nits were made in a more assertive fashion.

^_^ I guess it is more of how you want to collaborate with others, but I see these as "net positive"

stefanv replied:
Agreed, Mark! Per our contributor guide we do that for newcomers, and check with core devs first. FWIW you all can push corrections to all my branches! I can always revert them if I don't like them 😎

grlee77 replied:
Thank you for making the suggestions, @mkcor. From my end it is also fine to just push these kind of edits directly in the future.

mkcor replied:
> May I suggest that these stylistic nits at the end of a review be simply "applied" instead of suggested?

Agreed as well. I would have committed these and merged the PR today (Monday), had there been no activity in the meantime. I tend to leave things 'pending' (not blocking) over the weekend.

mkcor commented on skimage/measure/tests/test_moments.py:
```diff
@@ -13,6 +15,41 @@
+def compare_moments(m1, m2, thresh=1e-8):
+    """Compare two moments arrays.
+
+    Compares only values in the upper-left triangle of m1, m2 to account since
```
```suggestion
    Compares only values in the upper-left triangle of m1, m2 since
```
?

mkcor commented on skimage/measure/tests/test_moments.py:
```diff
@@ -13,6 +15,41 @@
+    values below the diagonal exceed the specified order and are not computed
+    when the analytical computation is used.
+
+    Also, there the first order central moments will be exactly zero with the
```
```suggestion
    Also, there the first-order central moments will be exactly zero with the
```
(optional)

mkcor reviewed (approved):
Thanks for wrapping this up, @grlee77!
